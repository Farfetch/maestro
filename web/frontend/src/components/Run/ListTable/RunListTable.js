import { Col, Row, Space, Table } from "antd";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";

import { runSingleUrl } from "../../../lib/routes";
import RunStatusTag from "../../tag/RunStatusTag";

const columns = [
  {
    title: "Title",
    dataIndex: "title",
    key: "title"
  },
  {
    title: "Notes",
    dataIndex: "notes",
    key: "notes"
  },
  {
    title: "Status",
    dataIndex: "runStatus",
    key: "runStatus",
    width: 180,
    render: (text) => <RunStatusTag runStatus={text} />
  },
  {
    title: "Started at",
    dataIndex: "createdAt",
    key: "createdAt",
    sorter: {
      compare: (recordA, recordB) => recordA.createdAt.diff(recordB.createdAt),
      multiple: 1
    },
    defaultSortOrder: "descend",
    render: (text, record) => record.createdAt.format("L HH:mm:ss"),
    width: 180
  },
  {
    title: "Action",
    key: "action",
    width: 100,
    render: (text, record) => (
      <Space size="middle">
        <Link to={runSingleUrl(record.key)}>Open</Link>
        {/* TODO: add delete to remove all data that is related to the test run */}
      </Space>
    )
  }
];

const runMapper = ({ id, title, runStatus, notes, createdAt }) => ({
  key: id,
  title,
  runStatus,
  notes,
  createdAt
});

const RunListTable = ({ runs, isLoading }) => {
  const runsDataSource = runs.map(runMapper);

  return (
    <Row gutter={[32, 32]} justify="start" align="middle">
      <Col span={24}>
        <Table
          size="small"
          loading={isLoading}
          dataSource={runsDataSource}
          columns={columns}
          pagination={{
            defaultPageSize: 50,
            hideOnSinglePage: true,
            pageSizeOptions: [50, 100, 200, 1000],
            position: ["bottomRight"]
          }}
        />
      </Col>
    </Row>
  );
};

RunListTable.propTypes = {
  runs: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      runStatus: PropTypes.string.isRequired,
      createdAt: PropTypes.object.isRequired,
      updatedAt: PropTypes.object.isRequired
    })
  ),
  isLoading: PropTypes.bool.isRequired
};

export default RunListTable;
