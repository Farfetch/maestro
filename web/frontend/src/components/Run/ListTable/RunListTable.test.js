import { render } from "@testing-library/react";
import moment from "moment";
import React from "react";
import { MemoryRouter } from "react-router-dom";

import RunListTable from "./RunListTable";

describe("components/RunConfiguration/Table", () => {
  const runs = [
    {
      id: "run-configuration-id",
      title: "Confgiuration title",
      runStatus: "RUNNING",
      notes: "Test notes field",
      createdAt: moment("2018-05-24T13:48:04.313000"),
      updatedAt: moment("2018-05-24T13:48:04.313000")
    }
  ];
  test(`should render RunConfigurationTable component`, async () => {
    const rendered = render(
      <MemoryRouter initialEntries={["/runs"]}>
        <RunListTable runs={runs} isLoading={false} />
      </MemoryRouter>
    );

    const badgeComponent = rendered.container;
    expect(badgeComponent.outerHTML).toMatchSnapshot();
  });

  test(`should render RunConfigurationTable loader`, async () => {
    const rendered = render(
      <MemoryRouter initialEntries={["/runs"]}>
        <RunListTable runs={runs} isLoading={true} />
      </MemoryRouter>
    );

    const badgeComponent = rendered.container;
    expect(badgeComponent.outerHTML).toMatchSnapshot();
  });
});
