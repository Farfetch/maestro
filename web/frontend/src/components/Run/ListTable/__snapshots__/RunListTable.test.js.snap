// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/RunConfiguration/Table should render RunConfigurationTable component 1`] = `"<div><div class=\\"ant-row ant-row-start ant-row-middle\\" style=\\"margin: -16px -16px -16px -16px;\\"><div style=\\"padding: 16px 16px 16px 16px;\\" class=\\"ant-col ant-col-24\\"><div class=\\"ant-table-wrapper\\"><div class=\\"ant-spin-nested-loading\\"><div class=\\"ant-spin-container\\"><div class=\\"ant-table ant-table-small\\"><div class=\\"ant-table-container\\"><div class=\\"ant-table-content\\"><table style=\\"table-layout: auto;\\"><colgroup><col><col><col style=\\"width: 180px;\\"><col style=\\"width: 180px;\\"><col style=\\"width: 100px;\\"></colgroup><thead class=\\"ant-table-thead\\"><tr><th class=\\"ant-table-cell\\">Title</th><th class=\\"ant-table-cell\\">Notes</th><th class=\\"ant-table-cell\\">Status</th><th class=\\"ant-table-cell ant-table-column-sort ant-table-column-has-sorters\\"><div class=\\"ant-table-column-sorters\\"><span class=\\"ant-table-column-title\\">Started at</span><span class=\\"ant-table-column-sorter ant-table-column-sorter-full\\"><span class=\\"ant-table-column-sorter-inner\\"><span role=\\"img\\" aria-label=\\"caret-up\\" class=\\"anticon anticon-caret-up ant-table-column-sorter-up\\"><svg viewBox=\\"0 0 1024 1024\\" focusable=\\"false\\" data-icon=\\"caret-up\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z\\"></path></svg></span><span role=\\"img\\" aria-label=\\"caret-down\\" class=\\"anticon anticon-caret-down ant-table-column-sorter-down active\\"><svg viewBox=\\"0 0 1024 1024\\" focusable=\\"false\\" data-icon=\\"caret-down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z\\"></path></svg></span></span></span></div></th><th class=\\"ant-table-cell\\">Action</th></tr></thead><tbody class=\\"ant-table-tbody\\"><tr data-row-key=\\"run-configuration-id\\" class=\\"ant-table-row ant-table-row-level-0\\"><td class=\\"ant-table-cell\\">Confgiuration title</td><td class=\\"ant-table-cell\\">Test notes field</td><td class=\\"ant-table-cell\\"><span class=\\"ant-tag ant-tag-processing\\"><span role=\\"img\\" aria-label=\\"sync\\" class=\\"anticon anticon-sync anticon-spin\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"sync\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M168 504.2c1-43.7 10-86.1 26.9-126 17.3-41 42.1-77.7 73.7-109.4S337 212.3 378 195c42.4-17.9 87.4-27 133.9-27s91.5 9.1 133.8 27A341.5 341.5 0 01755 268.8c9.9 9.9 19.2 20.4 27.8 31.4l-60.2 47a8 8 0 003 14.1l175.7 43c5 1.2 9.9-2.6 9.9-7.7l.8-180.9c0-6.7-7.7-10.5-12.9-6.3l-56.4 44.1C765.8 155.1 646.2 92 511.8 92 282.7 92 96.3 275.6 92 503.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8zm756 7.8h-60c-4.4 0-7.9 3.5-8 7.8-1 43.7-10 86.1-26.9 126-17.3 41-42.1 77.8-73.7 109.4A342.45 342.45 0 01512.1 856a342.24 342.24 0 01-243.2-100.8c-9.9-9.9-19.2-20.4-27.8-31.4l60.2-47a8 8 0 00-3-14.1l-175.7-43c-5-1.2-9.9 2.6-9.9 7.7l-.7 181c0 6.7 7.7 10.5 12.9 6.3l56.4-44.1C258.2 868.9 377.8 932 512.2 932c229.2 0 415.5-183.7 419.8-411.8a8 8 0 00-8-8.2z\\"></path></svg></span><span>RUNNING</span></span></td><td class=\\"ant-table-cell ant-table-column-sort\\">05/24/2018 13:48:04</td><td class=\\"ant-table-cell\\"><div class=\\"ant-space ant-space-horizontal ant-space-align-center\\"><div class=\\"ant-space-item\\"><a href=\\"/run/run-configuration-id\\">Open</a></div></div></td></tr></tbody></table></div></div></div></div></div></div></div></div></div>"`;

exports[`components/RunConfiguration/Table should render RunConfigurationTable loader 1`] = `"<div><div class=\\"ant-row ant-row-start ant-row-middle\\" style=\\"margin: -16px -16px -16px -16px;\\"><div style=\\"padding: 16px 16px 16px 16px;\\" class=\\"ant-col ant-col-24\\"><div class=\\"ant-table-wrapper\\"><div class=\\"ant-spin-nested-loading\\"><div><div class=\\"ant-spin ant-spin-spinning\\"><span class=\\"ant-spin-dot ant-spin-dot-spin\\"><i class=\\"ant-spin-dot-item\\"></i><i class=\\"ant-spin-dot-item\\"></i><i class=\\"ant-spin-dot-item\\"></i><i class=\\"ant-spin-dot-item\\"></i></span></div></div><div class=\\"ant-spin-container ant-spin-blur\\"><div class=\\"ant-table ant-table-small\\"><div class=\\"ant-table-container\\"><div class=\\"ant-table-content\\"><table style=\\"table-layout: auto;\\"><colgroup><col><col><col style=\\"width: 180px;\\"><col style=\\"width: 180px;\\"><col style=\\"width: 100px;\\"></colgroup><thead class=\\"ant-table-thead\\"><tr><th class=\\"ant-table-cell\\">Title</th><th class=\\"ant-table-cell\\">Notes</th><th class=\\"ant-table-cell\\">Status</th><th class=\\"ant-table-cell ant-table-column-sort ant-table-column-has-sorters\\"><div class=\\"ant-table-column-sorters\\"><span class=\\"ant-table-column-title\\">Started at</span><span class=\\"ant-table-column-sorter ant-table-column-sorter-full\\"><span class=\\"ant-table-column-sorter-inner\\"><span role=\\"img\\" aria-label=\\"caret-up\\" class=\\"anticon anticon-caret-up ant-table-column-sorter-up\\"><svg viewBox=\\"0 0 1024 1024\\" focusable=\\"false\\" data-icon=\\"caret-up\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z\\"></path></svg></span><span role=\\"img\\" aria-label=\\"caret-down\\" class=\\"anticon anticon-caret-down ant-table-column-sorter-down active\\"><svg viewBox=\\"0 0 1024 1024\\" focusable=\\"false\\" data-icon=\\"caret-down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z\\"></path></svg></span></span></span></div></th><th class=\\"ant-table-cell\\">Action</th></tr></thead><tbody class=\\"ant-table-tbody\\"><tr data-row-key=\\"run-configuration-id\\" class=\\"ant-table-row ant-table-row-level-0\\"><td class=\\"ant-table-cell\\">Confgiuration title</td><td class=\\"ant-table-cell\\">Test notes field</td><td class=\\"ant-table-cell\\"><span class=\\"ant-tag ant-tag-processing\\"><span role=\\"img\\" aria-label=\\"sync\\" class=\\"anticon anticon-sync anticon-spin\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"sync\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M168 504.2c1-43.7 10-86.1 26.9-126 17.3-41 42.1-77.7 73.7-109.4S337 212.3 378 195c42.4-17.9 87.4-27 133.9-27s91.5 9.1 133.8 27A341.5 341.5 0 01755 268.8c9.9 9.9 19.2 20.4 27.8 31.4l-60.2 47a8 8 0 003 14.1l175.7 43c5 1.2 9.9-2.6 9.9-7.7l.8-180.9c0-6.7-7.7-10.5-12.9-6.3l-56.4 44.1C765.8 155.1 646.2 92 511.8 92 282.7 92 96.3 275.6 92 503.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8zm756 7.8h-60c-4.4 0-7.9 3.5-8 7.8-1 43.7-10 86.1-26.9 126-17.3 41-42.1 77.8-73.7 109.4A342.45 342.45 0 01512.1 856a342.24 342.24 0 01-243.2-100.8c-9.9-9.9-19.2-20.4-27.8-31.4l60.2-47a8 8 0 00-3-14.1l-175.7-43c-5-1.2-9.9 2.6-9.9 7.7l-.7 181c0 6.7 7.7 10.5 12.9 6.3l56.4-44.1C258.2 868.9 377.8 932 512.2 932c229.2 0 415.5-183.7 419.8-411.8a8 8 0 00-8-8.2z\\"></path></svg></span><span>RUNNING</span></span></td><td class=\\"ant-table-cell ant-table-column-sort\\">05/24/2018 13:48:04</td><td class=\\"ant-table-cell\\"><div class=\\"ant-space ant-space-horizontal ant-space-align-center\\"><div class=\\"ant-space-item\\"><a href=\\"/run/run-configuration-id\\">Open</a></div></div></td></tr></tbody></table></div></div></div></div></div></div></div></div></div>"`;
